File: 백준/Platinum/5719. 거의 최단 경로/거의 최단 경로.py
# 거의 최단 경로
import sys
input = sys.stdin.readline
import heapq as hq
from collections import deque

def dijkstra():
    que = []
    hq.heappush(que, (0, S, S))

    while que:
        cost, cur, prev = hq.heappop(que)

        if dist[cur] < cost:
            continue
        
        if dist[cur] == float('inf'):
            dist[cur] = cost
            route[prev][cur] = 1
        else:
            route[prev][cur] = 1
            continue

        for n, nc in graph[cur]:

            if dist[n] < nc + cost:
                continue
            elif banned[cur][n]:
                continue
            else:
                hq.heappush(que, (nc+cost, n, cur))

while True:
    N, M = map(int,input().split())
    if (N == 0 and M == 0):
        break

    S, D = map(int,input().split())
    dist = [float('inf')] * N
    route = [[0] * N for _ in range(N)]
    banned = [[0] * N for _ in range(N)]
    
    graph = [[] for _ in range(N)]
    for i in range(M):
        U, V, P  = map(int,input().split())
        graph[U].append((V, P))
    
    dijkstra()
    short = dist[D]

    deq = deque([D])
    while deq:
        cur = deq.popleft()
        
        for i in range(N):
            if i == cur:
                continue
            
            if route[i][cur] and banned[i][cur] == 0:
                banned[i][cur] = 1
                deq.append(i)

    dist = [float('inf')] * N
    dijkstra()

    if dist[D] == float('inf') or short == dist[D]:
        print(-1)
    else:
        print(dist[D])

File: 백준/Platinum/5719. 거의 최단 경로/거의 최단 경로.java
import java.io.*;
import java.util.*;

public class Main {
    static int N; static int M;
    static int S; static int D;

    static int[] dist;
    static boolean[][] route;
    static boolean[][] banned;
    static ArrayList<ArrayList<int[]>> graph;

    static class Point {
        int loc;
        int cost;
        int prev;

        public Point(int loc, int cost, int prev) {
            this.loc = loc;
            this.cost = cost;
            this.prev = prev;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        while (true) {
            st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            M = Integer.parseInt(st.nextToken());

            if (N == 0 && M == 0) {
                break;
            }

            st = new StringTokenizer(br.readLine());
            S = Integer.parseInt(st.nextToken());
            D = Integer.parseInt(st.nextToken());

            dist = new int[N];
            Arrays.fill(dist, Integer.MAX_VALUE);

            route = new boolean[N][N];
            banned = new boolean[N][N];

            graph = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                graph.add(new ArrayList<>());
            }

            for (int i = 0; i < M; i++) {
                st = new StringTokenizer(br.readLine());
                int U = Integer.parseInt(st.nextToken());
                int V = Integer.parseInt(st.nextToken());
                int P = Integer.parseInt(st.nextToken());
                graph.get(U).add(new int[]{V, P});
            }

            dijkstra();
            int shortest = dist[D];

            LinkedList<Integer> deq = new LinkedList<>();
            deq.push(D);

            while (!deq.isEmpty()) {
                int cur = deq.pollFirst();

                for (int i = 0; i < N; i++) {
                    if (route[i][cur] && !banned[i][cur]) {
                        banned[i][cur] = true;
                        deq.add(i);
                    }
                }
            }

            dist = new int[N];
            Arrays.fill(dist, Integer.MAX_VALUE);
            dijkstra();

            if (dist[D] == Integer.MAX_VALUE || shortest == dist[D]) {
                System.out.println(-1);
            } else {
                System.out.println(dist[D]);
            }
        }

    }

    static void dijkstra() {
        PriorityQueue<Point> pq = new PriorityQueue<>((a, b) -> a.cost - b.cost);
        pq.offer(new Point(S, 0, S));

        while (!pq.isEmpty()) {
            Point cur = pq.poll();

            if (dist[cur.loc] < cur.cost) {
                continue;
            }

            route[cur.prev][cur.loc] = true;

            if (dist[cur.loc] == Integer.MAX_VALUE) {
                dist[cur.loc] = cur.cost;
            } else {
                continue;
            }

            for (int[] tmp : graph.get(cur.loc)) {
                int n = tmp[0];
                int nc = tmp[1];

                if (dist[n] < nc + cur.cost) {
                    continue;
                } else if (banned[cur.loc][n]) {
                    continue;
                } else {
                    pq.offer(new Point(n, nc+cur.cost, cur.loc));
                }
            }

        }

    }


}
